Comparison to external PPI, etc. networks
-----------------------------------------

If the co-expression network is determined due factors relating to underlying
gene regulatory networks, protein-protein interaction networks, etc., then we
might expect to see some of the edges for those networks recapitulated in our
final co-expression network.

To examine this, we will look at several highly-curated sources of human and
mouse network information:

1. IRefIndex (PPI)
2. ConsensusPathDB (PPI)

### IRefIndex

Using the [iRefR](http://irefindex.org/wiki/index.php?title=iRefR) package in
an external script (2015/92-load-irefindex), an edge list was generated for all
ENSEMBL gene IDs that could be mapped from source proteins.

Below, we will load this edge list in, convert it to an adjacency matrix, and
then compare it to the co-expression network adjacency matrix.

```{r load_irefindex_network, eval=TRUE}
# input filepath
if (CONFIG$host == 'H. sapiens') {
    iref_filepath <- file.path(Sys.getenv('REF'), 'irefindex', '14.0',
                            '9606.07042015.ensembl.csv.gz')
} else if (CONFIG$host == 'M. musculus') {
    iref_filepath <- file.path(Sys.getenv('REF'), 'irefindex', '14.0',
                            '10090.07042015.ensembl.csv.gz')
}

# load network
iref_edgelist <- read.csv(gzfile(iref_filepath))
g <- graph.data.frame(iref_edgelist, directed=FALSE)

# remove nodes and edges for genes that arent in the co-expression network
num_before <- length(V(g))
g <- delete.vertices(g, V(g)[!V(g)$name %in% gene_ids])
num_after <- length(V(g))
print(sprintf("Removed %d vertices from iRefIndex not found in co-expression network", 
              num_before - num_after))

# create a new matrix with the same dimensions as the co-expression network
m <- Matrix(0, nrow=length(gene_ids), ncol=length(gene_ids), sparse=TRUE)
colnames(m) <- gene_ids
rownames(m) <- gene_ids

g <- graph.union(g, graph.adjacency(m, mode='undirected'))

# convert to an adjacency matrix
iref_adjmatrix <- get.adjacency(g)

# reorder rows and columns
iref_adjmatrix <- iref_adjmatrix[gene_ids, gene_ids]

# use binary irefindex adjacency matrix to mask co-expression adjmatrix
#masked_adjmatrix <- ifelse(iref_adjmatrix, adjacency_matrix, 0)
masked_adjmatrix <- adjacency_matrix
masked_adjmatrix[!as.matrix(iref_adjmatrix)] <- 0
```

```{r irefindex_plots, eval=TRUE}
# sum of irefindex supported edge weights
iref_edge_sum <- sum(masked_adjmatrix)

print(sprintf("Edge weight support by iRefIndex PPIs: %0.2f",
              iref_edge_sum))
print(sprintf("Ratio of total co-expression network edge weights: %f",
              sum(masked_adjmatrix) / sum(adjacency_matrix)))
print(sprintf("Number of edges with PPI support: %d", sum(masked_adjmatrix > 0)))

# weight of supported edges
sum(masked_adjmatrix > 0.8)
sum(masked_adjmatrix > 0.5)
sum(masked_adjmatrix > 0.3)
sum(masked_adjmatrix > 0.1)

# Heatmap
if (CONFIG$include_plots) {
    # Create a binary version of the matrix and reorder to match previous heatmaps
    ppi_input <- masked_adjmatrix[heatmap_row_idx, heatmap_col_idx]
    ppi_input[ppi_input > 0] <- 1

    # make diagonal easy to see
    diag(ppi_input) <- 1

    heatmap.2(ppi_input, 
              Rowv=NA, Colv=NA,
              labRow=NA, labCol=NA, 
              hclustfun=flashClust, col=c('#000000', '#00FF00'), margin=c(7,7),
              key.xlab='iRefIndex supported co-expression network',
              trace='none',
              dendrogram='none',
              xlab='Gene', ylab='Gene',
              cexRow=2, cexCol=2, keysize=1,
              density.info='none', revC=TRUE)

    # clean-up plot input when finished
    rm(ppi_input)
}
```

```{r irefindex_cleanup}
rm(masked_adjmatrix)
gc()
```

Transcript and UTR length impact on expression
----------------------------------------------

```{r}
# gene utr lengths
utr5 <- fiveUTRsByTranscript(orgdb, use.names=TRUE)
utr5_lengths <- mean(width(ranges(utr5)))
utr5_lengths <- setNames(as.data.frame(utr5_lengths), 'utr5_len') %>%
    tibble::rownames_to_column('TXNAME')

utr3 <- threeUTRsByTranscript(orgdb, use.names=TRUE)
utr3_lengths <- mean(width(ranges(utr3)))
utr3_lengths <- setNames(as.data.frame(utr3_lengths), 'utr3_len') %>%
    tibble::rownames_to_column('TXNAME')

# ENSEMBL gene / RefSeq tx mapping
gene_tx <- AnnotationDbi::select(orgdb, keys=gene_info$gene_id,
                                 keytype='ENSEMBL', columns='TXNAME')

# Add tx length info
gene_tx <- merge(gene_tx, merge(utr5_lengths, utr3_lengths, by='TXNAME'),
                 by='TXNAME')

# UTR lengths at the gene level
utr_info <- gene_tx %>% 
    group_by(gene_id=ENSEMBL) %>%
    summarize(utr5_len=mean(utr5_len), utr3_len=mean(utr3_len))
```

### UTR length by stage and DE status

```{r utr_length_correlation_with_stage_dea, results='asis'}
# create lists to store 5' and 3' UTR length bar plots
utr5_plts <- list()
utr3_plts <- list()

# iterate over contrasts
for (stage in names(top_genes)) {
    x <- top_genes[[stage]]

    num_up <- sum(x$adj.P.Val < 0.05 & x$logFC > 0)
    num_down <- sum(x$adj.P.Val < 0.05 & x$logFC < 0)

    # skip contrasts with few/no up- and down-regulated genes
    if (num_up < 10 || num_down < 10) {
        next
    }

    # split into up- and down-regulated sets
    x$de_status <- 'Non-DE'
    x$de_status[x$adj.P.Val < 0.05 & x$logFC > 0] <- sprintf('Up-regulated (%d genes)', num_up)
    x$de_status[x$adj.P.Val < 0.05 & x$logFC < 0] <- sprintf('Down-regulated (%d genes)', num_down)

    cat(sprintf('\n#### %s (%d up / %d down)\n\n', stage, num_up, num_down))

    x <- merge(x, utr_info, by='gene_id') 

    # table output
    xx <- x %>% 
        group_by(de_status) %>%
        summarise(
            median_5utr_len=median(utr5_len, na.rm=TRUE) * 1.0,
            median_3utr_len=median(utr3_len, na.rm=TRUE) * 1.0
        )

    print(kable(xx))
    cat('\n')

    # add newline between de status and gene numbers in plot
    xx$de_status <- sub(' \\(', '\n\\(', xx$de_status)

    # bar plots for contrasts with differing UTR lengths
    if (abs(xx$median_5utr_len[3] - xx$median_5utr_len[1]) > 10) {    
        utr5_plts[[stage]] <- ggplot(xx, aes(de_status, y=median_5utr_len)) +
            geom_bar(stat = "identity") +
            labs(title=sprintf("%s hrs", gsub('_', ' ', stage))) +
            theme(axis.title.x=element_blank(),
                axis.title.y=element_blank(),
                plot.title=element_text(hjust=0.5))
    }

    if (abs(xx$median_3utr_len[3] - xx$median_3utr_len[1]) > 10) {    
        utr3_plts[[stage]] <- ggplot(xx, aes(de_status, y=median_3utr_len)) +
            geom_bar(stat = "identity") +
            labs(title=sprintf("%s hrs", gsub('_', ' ', stage))) +
            theme(axis.title.x=element_blank(),
                axis.title.y=element_blank(),
                plot.title=element_text(hjust=0.5))
    }
}
```

```{r utr5_length_by_de_status, results='asis'}
if (length(utr5_plts) > 0) {
    grid.arrange(grobs=utr5_plts, ncol=3,
                bottom='Differential Expression Status',
                left="Median 5' UTR length")
}
```

```{r utr3_length_by_de_status, results='asis'}
if (length(utr3_plts) > 0) {
    grid.arrange(grobs=utr3_plts, ncol=3,
                bottom='Differential Expression Status',
                left="Median 3' UTR length")
}
```

