Output
------

```{r create_output_tables}
# Create result data frame
result <- gene_info

# drop unneeded columns
keep_cols <- c('gene_id', 'description', 'chromosome',
               'strand', 'transcript_length')
result <- tbl_df(result[,keep_cols])

# add differential expression status for each gene in output
for (contrast in names(sig_genes)) {
    de_status <- result$gene_id %in% sig_genes[[contrast]]$gene_id
    result[paste('de', contrast, sep='_')] <- de_status
}
```

```{r save_output}
# Save raw and final version of the count table
raw_counts_unfiltered_outfile <- file.path(output_dir, sprintf('%s_counts_raw_all_genes.csv', 
                                                    CONFIG$analysis_name)) 
raw_counts_filtered_outfile <- file.path(output_dir, sprintf('%s_counts_raw_filtered.csv', 
                                                    CONFIG$analysis_name)) 
final_counts_outfile <- file.path(output_dir, sprintf('%s_counts_final.csv', 
                                                      CONFIG$analysis_name))

write.csv(count_table, file=raw_counts_unfiltered_outfile, quote=FALSE)
write.csv(exprs(de_counts$raw), file=raw_counts_filtered_outfile, quote=FALSE)
write.csv(exprs(de_counts$final), file=final_counts_outfile, quote=FALSE)

# Counts by condition (raw)
cond_counts <- avearrays(exprs(de_counts$raw), ID=CONFIG$samples$condition)
colnames(cond_counts) <- paste0('AveRawExpr_', colnames(cond_counts))

# Save differential expression tables
for (contrast in names(sig_genes)) {
    # Add fold-change and p-values
    x <- as.data.frame(top_genes[[contrast]] %>% select(gene_id, logFC, adj.P.Val))
    x$FC <- 2**(x$logFC)

    # Add average expression per condition
    x <- merge(x, cond_counts, by.x='gene_id', by.y='row.names')

    # combine with gene metadata
    output <- join(result, x, by='gene_id')

    # Remove unrelated de comparisons
    other_contrasts <- paste0('de_', names(sig_genes)[contrast != names(sig_genes)])
    output <- output[,! colnames(output) %in% other_contrasts]

    # Create a copy of the table with only the DE genes
    output_de <- output[output$gene_id  %in% sig_genes[[contrast]]$gene_id,]

    # Rename _vs_ to _over_ to be more clear
    contrast <- sub('_vs_', '_over_', contrast)

    # Save tables
    write.table(output, file=file.path(output_dir, sprintf('%s_all.tab', contrast)),
                sep='\t', row.names=FALSE)
    write.table(output_de, file=file.path(output_dir, sprintf('%s_diffexpr.tab', contrast)),
                sep='\t', row.names=FALSE)
}
```
