Secreted Proteins
-----------------

### SignalP secreted proteins

Using annotations download from TriTrypDB 8.1 (accessed on 2014/10/07).
SignalP 4.1 was used to determine which genes might correspond to secreted or
membrane proteins.

From the SignalP documentation:

- C-score (raw cleavage site score)
- S-score (signal peptide score)
- Y-score (combined cleavage site score)

>    signalp  will  then report the maximal C-, S-, and Y- scores, the mean S-score
>    in the interval between the N-terminal and the site with the maximal Y-score
>    and,  finally, the  D-score,  a  weighted average of the S-mean and Y-max
>    score.  The final, combined prediction of the presence of a signal peptide in
>    the sequence is  based  on  the  D- score.

The commands used to generate the below table are:

```
signalp -t euk -f short -c 70                               |\
    TriTrypDB-9.0_LmajorFriedlin_AnnotatedProtein.fasta     |\
    tail -n +2                                              |\
    sed '1s/^..//'                                          |\
    sed 's/ \+/,/g' > TriTrypDB-9.0_LmajorFriedlin_SignalP.csv
```

```{r load_secreted_proteins, results='asis'}
secreted_proteins <- tbl_df(read.csv(secreted_proteins))
colnames(secreted_proteins) <- c("gene_id", "Cmax", "Cmax_pos", "Ymax",
                                "Ymax_pos", "Smax", "Smax_pos", "Smean", "D",
                                "secreted", "Dmaxcut", "networks_used")

# Keep only the genes that may be secreted
secreted_proteins <- secreted_proteins %>% filter(secreted == 'Y')
```

```{r results='asis', include=CONFIG$verbose, echo=CONFIG$verbose}
if (CONFIG$include_tables) {
    xkable(secreted_proteins)
}
```

```{r secreted_proteins_by_module, results='asis'}
# secreted gene status
secreted_gene_status <- rep(0, nrow(network_counts$final))
secreted_gene_status[featureNames(network_counts$final) %in%
                                secreted_proteins$gene_id] <- 1

# create a data frame with the number of secreted proteins in each module
secreted_freqs <- tbl_df(data.frame(cbind(
                            gene=featureNames(network_counts$final),
                            module=module_colors,
                            secreted=as.logical(secreted_gene_status))))
secreted_freqs <- secreted_freqs %>% group_by(module) %>%
    summarize(num_genes=n(), num_secreted=sum(secreted == TRUE)) %>%
    mutate(ratio_secreted=num_secreted/num_genes) %>%
    arrange(desc(ratio_secreted))


# significance
total_secreted <- sum(secreted_gene_status)

secreted_pvalues <- c()

for (color in secreted_freqs$module) {
    # create contingency matrix
    x <- secreted_freqs %>% filter(module==color)
    contingency_matrix <- matrix(c(
          x$num_secreted,
          total_secreted - x$num_secreted,
          x$num_genes - x$num_secreted,
          num_genes - total_secreted - (x$num_genes - x$num_secreted)),
    nrow=2)

    # fisher's extact test
    fet_result <- fisher.test(contingency_matrix, alternative='greater')
    secreted_pvalues <- append(secreted_pvalues, fet_result$p.value)
}

secreted_freqs <- cbind(secreted_freqs, p.value=secreted_pvalues,
                       p.value.adj=p.adjust(secreted_pvalues, method='BH'))

# save and output
COEXPRESSION_NETWORK_RESULT$secreted_freqs <- secreted_freqs

out <- secreted_freqs %>% filter(p.value.adj <= 0.05) %>% arrange(p.value.adj)

if (nrow(out) > 0 && CONFIG$include_tables) {
    xkable(out)
}
```

```{r secreted_proteins}
# create vector of enrichment status
secreted_module_status <- rep(0, nrow(network_counts$final))
enriched_secreted_modules <- (secreted_freqs %>% filter(p.value.adj < 0.05))$module
enriched_secreted_genes <- result$gene_id[result$color %in% enriched_secreted_modules]
secreted_module_status[featureNames(network_counts$final) %in% enriched_secreted_genes] <- 1
secreted_colors <- labels2colors(secreted_module_status)

if (CONFIG$include_plots) {
    plotDendroAndColors(gene_tree,
                        cbind(module_colors, secreted_module_status),
                        groupLabels=c(sprintf("Modules (n=%s)", num_modules),
                                    "Secreted proteins"),
                        dendroLabels=FALSE,
                        cex.colorLabels=cex_color_labels,
                        cex.main=cex_main,
                        cex.axis=cex_axis,
                        cex.lab=cex_lab,
                        marAll=c(4,8,6,4),
                        guideHang=0.05)
}
```

Transmembrane Proteins
-----------------

Using TMHMM 2.0c configured for "short" output. From the TMHMM manual:

- "len=": the length of the protein sequence.
- "ExpAA=": The expected number of amino acids intransmembrane helices. If this
  number is larger than 18 it is very likely to be a transmembrane protein (OR
  have a signal peptide).
- "First60=": The expected number of amino acids in transmembrane helices in the
              first 60 amino acids of the protein. If this number more than a
              few, you should be warned that a predicted transmembrane helix in
              the N-term could be a signal peptide.
- "PredHel=": The number of predicted transmembrane helices by N-best.
- "Topology=": The topology predicted by N-best.

Command used:

```
tmhmm TriTrypDB-9.0_LmajorFriedlin_AnnotatedProtein.fasta     |\
    sed 's/\t/,/g' | sed 's/len=//g' | sed 's/ExpAA=//g'      |\
    sed 's/First60=//g' | sed 's/PredHel=//g'                 |\
    sed 's/Topology=//g' > TriTrypDB-9.0_LmajorFriedlin_TMHMM.csv
```

```{r load_membrane_proteins}
membrane_proteins <- tbl_df(read.csv(membrane_proteins))
colnames(membrane_proteins) <- c('gene_id', 'len', 'exp_aa', 'first60',
                                'pred_hel', 'toplogy')
# Keep only the genes that have at least one transmembrane domain and are
# not likely to be secreted
membrane_proteins <- membrane_proteins %>% filter(pred_hel > 0 & first60 < 3)

# Create vector of gene statuses
membrane_gene_status <- rep(0, nrow(network_counts$final))
membrane_gene_status[featureNames(network_counts$final) %in% membrane_proteins$gene_id] <- 1
```

```{r results='asis', include=CONFIG$verbose, echo=CONFIG$verbose}
if (CONFIG$include_tables) {
    xkable(membrane_proteins)
}
```

```{r membrane_proteins_by_module, results='asis'}
# create a data frame with the number of membrane proteins in each module
membrane_freqs <- tbl_df(data.frame(cbind(
                            gene=featureNames(network_counts$final),
                            module=module_colors,
                            membrane=as.logical(membrane_gene_status))))
membrane_freqs <- membrane_freqs %>% group_by(module) %>%
    summarize(num_genes=n(), num_membrane=sum(membrane == TRUE)) %>%
    mutate(ratio_membrane=num_membrane/num_genes) %>%
    arrange(desc(ratio_membrane))

# significance
total_membrane <- sum(membrane_gene_status)

membrane_pvalues <- c()

for (color in membrane_freqs$module) {
    # create contingency matrix
    x <- membrane_freqs %>% filter(module==color)
    contingency_matrix <- matrix(c(
          x$num_membrane,
          total_membrane - x$num_membrane,
          x$num_genes - x$num_membrane,
          num_genes - total_membrane - (x$num_genes - x$num_membrane)),
    nrow=2)

    # fisher's extact test
    fet_result <- fisher.test(contingency_matrix, alternative='greater')

    membrane_pvalues <- append(membrane_pvalues, fet_result$p.value)

}

membrane_freqs <- cbind(membrane_freqs, p.value=membrane_pvalues,
                       p.value.adj=p.adjust(membrane_pvalues, method='BH'))

# save and output
COEXPRESSION_NETWORK_RESULT$membrane_freqs <- membrane_freqs

if (CONFIG$include_tables) {
    xkable(membrane_freqs %>% filter(p.value.adj <= 0.05) %>% arrange(p.value.adj))
}
```

```{r membrane_proteins}
# create vector of enrichment status
membrane_module_status <- rep(0, nrow(network_counts$final))
enriched_tm_modules <- (membrane_freqs %>% filter(p.value.adj < 0.05))$module
enriched_tm_genes <- result$gene_id[result$color %in% enriched_tm_modules]
membrane_module_status[featureNames(network_counts$final) %in% enriched_tm_genes] <- 1
membrane_colors <- labels2colors(membrane_module_status)

if (CONFIG$include_plots) {
    plotDendroAndColors(gene_tree,
                        cbind(module_colors, membrane_module_status),
                        groupLabels=c(sprintf("Modules (n=%s)", num_modules),
                                    "Membrane proteins"),
                        dendroLabels=FALSE,
                        cex.colorLabels=cex_color_labels,
                        cex.main=cex_main,
                        cex.axis=cex_axis,
                        cex.lab=cex_lab,
                        marAll=c(4,8,6,4),
                        guideHang=0.05)
}
```

GPI-anchored Proteins
---------------------

To predict GPI-anchored proteins, the [PredGPI web
server](http://gpcr.biocomp.unibo.it/predgpi/pred.htm) was used.

TriTrypDB XXAnnotatedProteins.fasta files were downloaded for each species and
split into smaller files with less than 500 entries each using the
[fasta-splitter.pl](http://kirill-kryukov.com/study/tools/fasta-splitter/)
script written by [Kirill Kryukov](http://kirill-kryukov.com):

```
fasta-splitter.pl --n-parts 25  TriTrypDB-24_LmajorFriedlin_AnnotatedProteins.fasta
```

Next, the divided files were submitted to the PredGPI server, and results were
downloaded and renamed to have a matching filename as the input FASTAs, with
the extension changed to ".txt" (e.g.
"TriTrypDB-24_TcruziCLBrenerEsmeraldo-like_AnnotatedProteins.part-01.txt")

Next, the output was parsed and combined into a single CSV using csvkit and
other standard linux tools:

```
grep -h ">" *.txt | csvcut -d '|' -c 1 | cut -c2- | tr -d ' ' > ids
grep -h ">" *.txt | csvcut -d '|' -c 8 | cut -c9- > vals
echo "gene_id,fdr" > TriTrypDB-24_LmajorFriedlin_PredGPI.csv
paste -d "," ids vals >> TriTrypDB-24_LmajorFriedlin_PredGPI.csv
```

```{r load_predicted_gpi_anchored_proteins}
# T. cruzi
if (CONFIG$pathogen != 'L. major') {
    # Load PredGPI predicted GPI-anchored gene ids
    gpi_anchored_df <- read_csv(gpi_anchored_proteins)
    gpi_anchored_df$fdr <- as.numeric(gpi_anchored_df$fdr)
                               #col_types=c(col_character(), col_double()))

    gpi_anchored <- (gpi_anchored_df %>% filter(fdr == 0))$gene_id
}

# L. major
# Based off of table 9.1 in "Leishmania After The Genome" (Myler et al, 2009)
if (CONFIG$pathogen == "L. major") {
    gpi_anchored <- c(
        'LmjF.12.0810', 'LmjF.12.1000', 'LmjF.12.0980', 'LmjF.12.0940',
        'LmjF.12.0780', 'LmjF.12.0730', 'LmjF.12.0890', 'LmjF.12.0990',
        'LmjF.12.0760', 'LmjF.10.0480', 'LmjF.10.0460', 'LmjF.10.0465',
        'LmjF.12.0850', 'LmjF.12.0830', 'LmjF.12.0960', 'LmjF.12.0740',
        'LmjF.12.1070', 'LmjF.12.0755', 'LmjF.12.0765', 'LmjF.12.0870',
        'LmjF.35.0290', 'LmjF.12.1005', 'LmjF.05.0900', 'LmjF.04.0140',
        'LmjF.04.0160', 'LmjF.04.0150', 'LmjF.04.0170', 'LmjF.04.0130',
        'LmjF.04.0210', 'LmjF.28.0570', 'LmjF.04.0180', 'LmjF.31.1440',
        'LmjF.15.0570', 'LmjF.04.0190', 'LmjF.09.0580', 'LmjF.31.2000',
        'LmjF.35.5240', 'LmjF.19.0170', 'LmjF.04.0200', 'LmjF.23.1170',
        'LmjF.09.0890', 'LmjF.34.2200', 'LmjF.34.2650', 'LmjF.34.3630',
        'LmjF.14.0770', 'LmjF.32.3720', 'LmjF.08.1080', 'LmjF.08.1030',
        'LmjF.08.1060', 'LmjF.13.0480', 'LmjF.06.1120', 'LmjF.20.0360',
        'LmjF.27.0070', 'LmjF.12.0400', 'LmjF.35.3220', 'LmjF.01.0640',
        'LmjF.30.0805', 'LmjF.03.0470', 'LmjF.30.2990', 'LmjF.10.0540',
        'LmjF.10.0170')
    #gpi_anchored <- unique(c(gpi_anchored, lmajor_gpihmm_predicted))
}

# gpi gene status
gpi_anchored_gene_status <- rep(0, length(gene_ids))
gpi_anchored_gene_status[gene_ids %in% gpi_anchored] <- 1

# create a data frame with the number of gpi-anchored proteins in each module
gpi_anchored_freqs <- tbl_df(data.frame(cbind(
                            gene=gene_ids,
                            module=module_colors,
                            gpi=as.logical(gpi_anchored_gene_status))))
gpi_anchored_freqs <- gpi_anchored_freqs %>% group_by(module) %>%
    summarize(num_genes=n(), num_gpi=sum(gpi == TRUE)) %>%
    mutate(ratio_gpi=num_gpi/num_genes) %>%
    arrange(desc(ratio_gpi))

# significance
total_gpi <- sum(gpi_anchored_gene_status)

gpi_anchored_pvalues <- c()

for (color in gpi_anchored_freqs$module) {
    # create contingency matrix
    #
    # ------------------------------------------
    # |    module gpi     |   module not gpi   |
    # |-------------------|---------------------
    # | not module gpi    | not module not cpi |
    # ------------------------------------------
    #
    x <- gpi_anchored_freqs %>% filter(module==color)
    contingency_matrix <- matrix(c(
          x$num_gpi,
          total_gpi - x$num_gpi,
          x$num_genes - x$num_gpi,
          num_genes - total_gpi - (x$num_genes - x$num_gpi)),
    nrow=2)

    # fisher's extact test
    fet_result <- fisher.test(contingency_matrix, alternative='greater')
    gpi_anchored_pvalues <- append(gpi_anchored_pvalues, fet_result$p.value)
}

gpi_anchored_freqs <- cbind(gpi_anchored_freqs, p.value=gpi_anchored_pvalues,
                       p.value.adj=p.adjust(gpi_anchored_pvalues, method='BH'))

# save and output
COEXPRESSION_NETWORK_RESULT$gpi_anchored_freqs <- gpi_anchored_freqs

if (CONFIG$include_tables) {
    xkable(gpi_anchored_freqs %>% filter(p.value.adj <= 0.05) %>% arrange(p.value.adj))
}
```

```{r gpi_anchored_proteins_output}
# create vector of enrichment status
gpi_anchored_module_status <- rep(0, nrow(network_counts$final))
enriched_tm_modules <- (gpi_anchored_freqs %>% filter(p.value.adj < 0.05))$module
enriched_tm_genes <- result$gene_id[result$color %in% enriched_tm_modules]
gpi_anchored_module_status[featureNames(network_counts$final) %in% enriched_tm_genes] <- 1
gpi_anchored_colors <- labels2colors(gpi_anchored_module_status)

if (CONFIG$include_plots) {
    plotDendroAndColors(gene_tree,
                        cbind(module_colors, gpi_anchored_module_status),
                        groupLabels=c(sprintf("Modules (n=%s)", num_modules),
                                    "GPI-anchored proteins"),
                        dendroLabels=FALSE,
                        cex.colorLabels=cex_color_labels,
                        cex.main=cex_main,
                        cex.axis=cex_axis,
                        cex.lab=cex_lab,
                        marAll=c(4,8,6,4),
                        guideHang=0.05)
}
```

Secreted/TM/GPI-anchored protein overlap
----------------------------------------

Question: How much do the predicted secreted, transmembrane and GPI-anchored
gene ids overlap?

```{r secreted_tm_overlap}
#secreted_genes <- featureNames(network_counts$final)[secreted_status == 1]
#membrane_genes <- featureNames(network_counts$final)[membrane_status == 1]
num_shared_st <- length(intersect(secreted_proteins$gene_id,
                                 membrane_proteins$gene_id))
num_shared_sg <- length(intersect(secreted_proteins$gene_id,
                                 gpi_anchored))
num_shared_tg <- length(intersect(membrane_proteins$gene_id,
                                 gpi_anchored))
num_shared_all <- length(Reduce(intersect,
                               list(membrane_proteins$gene_id,
                                    secreted_proteins$gene_id,
                                    gpi_anchored)))

if (CONFIG$include_plots && 'venneuler' %in% installed.packages()) {
    require(venneuler)
    plot(venneuler(c(
        Secreted=nrow(secreted_proteins),
        Transmembrane=nrow(membrane_proteins),
        GPI=length(gpi_anchored),
        "Secreted&Transmembrane"=num_shared_st,
        "Secreted&GPI"=num_shared_sg,
        "Transmembrane&GPI"=num_shared_tg,
        "Secreted&Transmembrane&GPI"=num_shared_all
    )))
}

print(sprintf("Number secreted genes: %d", nrow(secreted_proteins)))
print(sprintf("Number transmembrane genes: %d", nrow(membrane_proteins)))
print(sprintf("Number GPI-anchored genes: %d", length(gpi_anchored)))
print(sprintf("Number overlapping genes: %d", num_shared_all))
```

