```{r knitr_settings, include=FALSE, message=FALSE, warning=FALSE}
library('knitr')
opts_knit$set(progress=FALSE, verbose=TRUE)
opts_chunk$set(error=FALSE)
options(digits=4)
options(stringsAsFactors=FALSE)
options(knitr.duplicate.label='allow')

# If rmarkdown.pandoc.to not specified (for example, when kniting
# piece-by-piece in Vim-R), have it default to 'latex' output.
if (is.null(opts_knit$get("rmarkdown.pandoc.to"))) {
    opts_knit$set(rmarkdown.pandoc.to='latex')
}

# Print bibliography as markdown
options("citation_format"="pandoc")

rm(list=ls())    # Clean up any existing variables
```

<!--

The goal of this script is to explore the use of [Weighted Gene Co-expression
Network Analysis (WGCNA)](http://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/)
to construct co-expression networks for a range of host and pathogen species, 
and to combine external sources of information in order to interrogate the
resulting networks.

The basic process begins by reading microarray or RNA-Seq expression data along
with gene annotation data (location, description, etc). Next, data is
normalized, filtered, and optionally adjusted for batch effects and tables and
plots (PCA, MA, heatmaps, etc) are generated to inspect the data.

Next, differential expression analysis is performed using Voom and Limma, in
order to provide additional insight into the important genes in the network.

WGCNA is then used to construct inspect the data, remove any sample outliers
that were not already detected during previous steps, and build a co-expression
network, and to detect network modules.

Once the network has been constructed, the nodes are annotated with additional
data from various sources (3'UTR motifs, DE enrichment, etc) and exported
into several formats that can be used for downstream visualization and
analysis.

Throughout the process various steps in the analysis can be controled via
settings in the `CONFIG` variable below as described below.

-->

Libraries
---------

Let's start by loading all of the libraries that will be used throughout this
analysis.

```{r random_seed}
# Make sure results are reproducible
set.seed(1)
```

```{r child='includes_common.Rmd'}
```

```{r child='includes_wgcna.Rmd'}
```

Parameters
----------

In order to simplify the process of rerunning the following analysis with
different parameters, I have tried to abstract out many of the useful
experimental design parameters, enabling them to all be adjusting in one
location.

The parameters are broken up into two parts: first are the general settings
which apply to all analyses, regardless of species. The second set of settings
are those which are specific to the species being analysed.

```{r child='../settings/shared.Rmd'}
```

```{r child='../settings/differential_expression.Rmd'}
```

```{r child='../settings/coex_network.Rmd'}
```
